### PR TITLE
hevm gst compliance

### PR DESCRIPTION
**Pull request by rainbreak**
`call` has different write semantics to other operations, which generally pad with zeros up to any excess of size + offset over the data length. With `call`, `writeMemory`was erroneously truncating memory in the case of calls where the bytes to write were shorter than the memory size given. We fix this with a special write method for `call`.

A number of other memory writing issues were fixed:

- correct zero padding of non-call memory operations
- `returndatacopy` is meant to fail if reading from beyond what has been written.
- memory offset calculations are not modulo-256 in some cases

Lastly, test failures due solely to incorrect nonce or storage were separated and counted.

Previously:

```
passed: 5868
bad-balance: 3948
bad-nonce: 0
bad-storage: 0
failed: 245
skipped: 1923
```

Now (up to 5e67834c798a813f2b987be6bf955080977208d5):

```
passed: 5887
bad-balance: 3944
bad-nonce: 16
bad-storage: 14
failed: 200
skipped: 1923
```

The `run-blockchain-test`was modified, with `--test` and `--skip` arguments implemented for finer control over tests that are run. This is used here for modexp, which typically goes out of memory. 

```patch
-BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callcodeOutput3_d0g0v0.json callcodeOutput3_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callcodeOutput3_d0g0v0.json callcodeOutput3_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callcodeOutput3partial_d0g0v0.json callcodeOutput3partial_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callcodeOutput3partial_d0g0v0.json callcodeOutput3partial_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callOutput3_d0g0v0.json callOutput3_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callOutput3_d0g0v0.json callOutput3_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callOutput3partial_d0g0v0.json callOutput3partial_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/callOutput3partial_d0g0v0.json callOutput3partial_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/createNameRegistratorPreStore1NotEnoughGas_d0g0v0.json createNameRegistratorPreStore1NotEnoughGas_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCallCreateCallCodeTest/createNameRegistratorPreStore1NotEnoughGas_d0g0v0.json createNameRegistratorPreStore1NotEnoughGas_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stCreate2/Create2OnDepth1024_d0g0v0.json Create2OnDepth1024_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreate2/Create2OnDepth1024_d0g0v0.json Create2OnDepth1024_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stCreate2/Create2OOGafterInitCodeReturndata_d0g1v0.json Create2OOGafterInitCodeReturndata_d0g1v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreate2/Create2OOGafterInitCodeReturndata_d0g1v0.json Create2OOGafterInitCodeReturndata_d0g1v0_ConstantinopleFix bad-balance

-BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_afterFailing_create_d0g0v0.json returndatacopy_afterFailing_create_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_afterFailing_create_d0g0v0.json returndatacopy_afterFailing_create_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_create_d1g0v0.json returndatacopy_following_create_d1g0v0_ConstantinopleFix bad-balance
-BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_revert_in_create_d0g0v0.json returndatacopy_following_revert_in_create_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_successful_create_d0g0v0.json returndatacopy_following_successful_create_d0g0v0_ConstantinopleFix bad-balance
+BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_create_d1g0v0.json returndatacopy_following_create_d1g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_revert_in_create_d0g0v0.json returndatacopy_following_revert_in_create_d0g0v0_ConstantinopleFix bad-nonce
+BlockchainTests/GeneralStateTests/stCreate2/returndatacopy_following_successful_create_d0g0v0.json returndatacopy_following_successful_create_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stCreate2/RevertOpcodeCreate_d0g0v0.json RevertOpcodeCreate_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreate2/RevertOpcodeCreate_d0g0v0.json RevertOpcodeCreate_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stCreate2/RevertOpcodeInCreateReturnsCreate2_d0g0v0.json RevertOpcodeInCreateReturnsCreate2_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreate2/RevertOpcodeInCreateReturnsCreate2_d0g0v0.json RevertOpcodeInCreateReturnsCreate2_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stCreateTest/CreateOOGafterInitCodeReturndata_d0g1v0.json CreateOOGafterInitCodeReturndata_d0g1v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stCreateTest/CreateOOGafterInitCodeReturndata_d0g1v0.json CreateOOGafterInitCodeReturndata_d0g1v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callcodeOutput3_d0g0v0.json callcodeOutput3_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callcodeOutput3_d0g0v0.json callcodeOutput3_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callOutput3_d0g0v0.json callOutput3_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callOutput3partial_d0g0v0.json callOutput3partial_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callOutput3_d0g0v0.json callOutput3_d0g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stDelegatecallTestHomestead/callOutput3partial_d0g0v0.json callOutput3partial_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallCodeGasMemoryAsk_d0g0v0.json RawCallCodeGasMemoryAsk_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallCodeGasMemoryAsk_d0g0v0.json RawCallCodeGasMemoryAsk_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallCodeGasValueTransferMemoryAsk_d0g0v0.json RawCallCodeGasValueTransferMemoryAsk_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallCodeGasValueTransferMemoryAsk_d0g0v0.json RawCallCodeGasValueTransferMemoryAsk_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallGasValueTransferMemoryAsk_d0g0v0.json RawCallGasValueTransferMemoryAsk_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallGasValueTransferMemoryAsk_d0g0v0.json RawCallGasValueTransferMemoryAsk_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallMemoryGasAsk_d0g0v0.json RawCallMemoryGasAsk_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawCallMemoryGasAsk_d0g0v0.json RawCallMemoryGasAsk_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawDelegateCallGasMemoryAsk_d0g0v0.json RawDelegateCallGasMemoryAsk_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stEIP150singleCodeGasPrices/RawDelegateCallGasMemoryAsk_d0g0v0.json RawDelegateCallGasMemoryAsk_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stHomesteadSpecific/contractCreationOOGdontLeaveEmptyContract_d0g0v0.json contractCreationOOGdontLeaveEmptyContract_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stHomesteadSpecific/contractCreationOOGdontLeaveEmptyContract_d0g0v0.json contractCreationOOGdontLeaveEmptyContract_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stHomesteadSpecific/createContractViaContractOOGInitCode_d0g0v0.json createContractViaContractOOGInitCode_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stHomesteadSpecific/createContractViaContractOOGInitCode_d0g0v0.json createContractViaContractOOGInitCode_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stInitCodeTest/CallRecursiveContract_d0g0v0.json CallRecursiveContract_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stInitCodeTest/CallRecursiveContract_d0g0v0.json CallRecursiveContract_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stInitCodeTest/OutOfGasPrefundedContractCreation_d0g3v0.json OutOfGasPrefundedContractCreation_d0g3v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stInitCodeTest/OutOfGasPrefundedContractCreation_d0g3v0.json OutOfGasPrefundedContractCreation_d0g3v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stMemExpandingEIP150Calls/CallAskMoreGasOnDepth2ThenTransactionHasWithMemExpandingCalls_d0g0v0.json CallAskMoreGasOnDepth2ThenTransactionHasWithMemExpandingCalls_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stMemExpandingEIP150Calls/CallAskMoreGasOnDepth2ThenTransactionHasWithMemExpandingCalls_d0g0v0.json CallAskMoreGasOnDepth2ThenTransactionHasWithMemExpandingCalls_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stQuadraticComplexityTest/Return50000_2_d0g1v0.json Return50000_2_d0g1v0_ConstantinopleFix bad-balance
+BlockchainTests/GeneralStateTests/stQuadraticComplexityTest/Return50000_2_d0g1v0.json Return50000_2_d0g1v0_ConstantinopleFix timeout

-BlockchainTests/GeneralStateTests/stRandom/randomStatetest85_d0g0v0.json randomStatetest85_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRandom/randomStatetest85_d0g0v0.json randomStatetest85_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stRecursiveCreate/recursiveCreate_d0g0v0.json recursiveCreate_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stRecursiveCreate/recursiveCreateReturnValue_d0g0v0.json recursiveCreateReturnValue_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRecursiveCreate/recursiveCreate_d0g0v0.json recursiveCreate_d0g0v0_ConstantinopleFix bad-nonce
+BlockchainTests/GeneralStateTests/stRecursiveCreate/recursiveCreateReturnValue_d0g0v0.json recursiveCreateReturnValue_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_after_failing_callcode_d0g0v0.json returndatacopy_after_failing_callcode_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_afterFailing_create_d0g0v0.json returndatacopy_afterFailing_create_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_after_failing_callcode_d0g0v0.json returndatacopy_after_failing_callcode_d0g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_afterFailing_create_d0g0v0.json returndatacopy_afterFailing_create_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_after_failing_staticcall_d0g0v0.json returndatacopy_after_failing_staticcall_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_after_failing_staticcall_d0g0v0.json returndatacopy_after_failing_staticcall_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_create_d0g0v0.json returndatacopy_following_create_d0g0v0_ConstantinopleFix bad-balance
-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_failing_call_d0g0v0.json returndatacopy_following_failing_call_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_create_d0g0v0.json returndatacopy_following_create_d0g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_failing_call_d0g0v0.json returndatacopy_following_failing_call_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_revert_in_create_d0g0v0.json returndatacopy_following_revert_in_create_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_successful_create_d0g0v0.json returndatacopy_following_successful_create_d0g0v0_ConstantinopleFix bad-balance
-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_too_big_transfer_d0g0v0.json returndatacopy_following_too_big_transfer_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_revert_in_create_d0g0v0.json returndatacopy_following_revert_in_create_d0g0v0_ConstantinopleFix bad-nonce
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_successful_create_d0g0v0.json returndatacopy_following_successful_create_d0g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_following_too_big_transfer_d0g0v0.json returndatacopy_following_too_big_transfer_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_initial_256_d2g0v0.json returndatacopy_initial_256_d2g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_initial_256_d2g0v0.json returndatacopy_initial_256_d2g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_initial_d0g0v0.json returndatacopy_initial_d0g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_overrun_d0g0v0.json returndatacopy_overrun_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_initial_d0g0v0.json returndatacopy_initial_d0g0v0_ConstantinopleFix ok
+BlockchainTests/GeneralStateTests/stReturnDataTest/returndatacopy_overrun_d0g0v0.json returndatacopy_overrun_d0g0v0_ConstantinopleFix ok

-BlockchainTests/GeneralStateTests/stRevertTest/RevertOpcodeCreate_d0g0v0.json RevertOpcodeCreate_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRevertTest/RevertOpcodeCreate_d0g0v0.json RevertOpcodeCreate_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stRevertTest/RevertOpcodeInCreateReturns_d0g0v0.json RevertOpcodeInCreateReturns_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRevertTest/RevertOpcodeInCreateReturns_d0g0v0.json RevertOpcodeInCreateReturns_d0g0v0_ConstantinopleFix bad-nonce

-BlockchainTests/GeneralStateTests/stRevertTest/RevertPrecompiledTouch_storage_d0g0v0.json RevertPrecompiledTouch_storage_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRevertTest/RevertPrecompiledTouch_storage_d0g0v0.json RevertPrecompiledTouch_storage_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stRevertTest/RevertPrecompiledTouch_storage_d3g0v0.json RevertPrecompiledTouch_storage_d3g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stRevertTest/RevertPrecompiledTouch_storage_d3g0v0.json RevertPrecompiledTouch_storage_d3g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stSpecialTest/FailedCreateRevertsDeletion_d0g0v0.json FailedCreateRevertsDeletion_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stSpecialTest/FailedCreateRevertsDeletion_d0g0v0.json FailedCreateRevertsDeletion_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stSStoreTest/InitCollision_d1g0v0.json InitCollision_d1g0v0_ConstantinopleFix bad-state
-BlockchainTests/GeneralStateTests/stSStoreTest/InitCollision_d2g0v0.json InitCollision_d2g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stSStoreTest/InitCollision_d1g0v0.json InitCollision_d1g0v0_ConstantinopleFix bad-storage
+BlockchainTests/GeneralStateTests/stSStoreTest/InitCollision_d2g0v0.json InitCollision_d2g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stZeroCallsTest/ZeroValue_TransactionCALL_ToOneStorageKey_d0g0v0.json ZeroValue_TransactionCALL_ToOneStorageKey_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stZeroCallsTest/ZeroValue_TransactionCALL_ToOneStorageKey_d0g0v0.json ZeroValue_TransactionCALL_ToOneStorageKey_d0g0v0_ConstantinopleFix bad-storage

-BlockchainTests/GeneralStateTests/stZeroCallsTest/ZeroValue_TransactionCALLwithData_ToOneStorageKey_d0g0v0.json ZeroValue_TransactionCALLwithData_ToOneStorageKey_d0g0v0_ConstantinopleFix bad-state
+BlockchainTests/GeneralStateTests/stZeroCallsTest/ZeroValue_TransactionCALLwithData_ToOneStorageKey_d0g0v0.json ZeroValue_TransactionCALLwithData_ToOneStorageKey_d0g0v0_ConstantinopleFix bad-storage

-passed: 5868
-bad-balance: 3948
-bad-nonce: 0
-bad-storage: 0
-failed: 245
+passed: 5887
+bad-balance: 3944
+bad-nonce: 16
+bad-storage: 14
+failed: 200
```

### CODE REVIEW

File: src/hevm/src/EVM/Concrete.hs [hunk 1]
```diff
@@ -98,21 +101,19 @@ sliceMemory o s m =
...
-      b'     = BS.drop (num n) b
-    in
-      a <> BS.replicate (num dst - BS.length a) 0 <> c <> b'
+  let
+    (a, b) = BS.splitAt (num dst) bs0
+    a'     = BS.replicate (num dst - BS.length a) 0
+    -- sliceMemory should work for both cases, but we are using 256 bit
+    -- words, whereas ByteString is only defined up to 64 bit. For large n,
+    -- src, dst this will cause problems (often in GeneralStateTests).
+    -- Later we could reimplement ByteString for 256 bit arguments.
```
Comment (rainbreak) [inline src/hevm/src/EVM/Concrete.hs]: @livnev ^ n.b. Data.ByteString limitations with large numbers. In practice the size limit was somewhat higher than int64, but well below 256.

This gives an interesting "feature" of Data.ByteString is that `bs == BS.drop n bs` for sufficiently large `n`.

**Comment by rainbreak** [comment]
With the last two commits, coverage increases to

```
passed: 5918
bad-balance: 3912
bad-nonce: 16
bad-storage: 14
failed: 201
skipped: 1923
```

**Comment by rainbreak** [comment]
@MrChico want to append your stuff in #247 to this branch? You can then run the tests with

```
./run-blockchain-tests tests --skip modexp
```

Where `tests` is the location of your `ethereum/tests` repo. The html is on stdout and there is simpler info on stderr.

**Comment by rainbreak** [comment]
The commit messages contain more results, but the current state is

    passed: 6705
    bad-balance: 3188
    bad-nonce: 17
    bad-storage: 8
    failed: 143
    skipped: 1923

File: src/hevm/src/EVM.hs [hunk 2]
```diff
@@ -2085,6 +2142,17 @@ costOfCall (FeeSchedule {..}) recipientExists xValue availableGas xGas =
       if availableGas >= c_extra
       then min xGas (allButOne64th (availableGas - c_extra))
       else xGas
+    c_foo = trace ("\n" <> " c_new:     " <> show c_new  <> "\n"
```
Comment (MrChico) [inline src/hevm/src/EVM.hs]: is this supposed to be here? Or is this debugging material sneaking into production

Comment (rainbreak) [reply src/hevm/src/EVM.hs]: oops yes, will remove

**Comment by rainbreak** [comment]
The refund / self destruct tracking from @mrchico in fe5c623 gave a huge improvement. The current status, with `--skip modexp`, is

```
passed: 9773
bad-balance: 221
bad-nonce: 1
bad-storage: 8
failed: 58
skipped: 1923
```

**Comment by rainbreak** [comment]
https://docs.google.com/spreadsheets/u/1/d/1xat7UI8GtB4ZGVdlK5_XQSHJZaMThi4SrlcL8XMZb5Q/htmlview#

Useful for context on remaining cases.

**Comment by rainbreak** [comment]
Now passing all the tests :) (using ethereum/tests@b8c9b4a6c, same as evm-semantics).

**Comment by MrChico** [comment]
Wowowow! Let's bump the version!

**Comment by livnev** [comment]
OMG! I can't believe it! amaaazzing

**Comment by livnev** [comment]
@rainbreak do you want to add `run-blockchain-tests` to the tests that CI will run for `dapptools`, to protect against regressions?

**Comment by rainbreak** [comment]
Good idea, though I wonder how Travis will cope with some of the more demanding tests like the above timeout.

**Comment by desaperados** [comment]
WOW.

**Comment by rainbreak** [comment]
TODO:

- [ ] add run-blockchain-tests to hevm.cabal data files
- [ ] bump the ethereum/tests commit in release.nix
- [ ] add something like hevm vm-test-report, but for blockchain-tests
- [ ] consider adding --html flag to run-blockchain-tests to control html output or not
- [ ] run the tests on travis build and reject build if failing (skipping known failures)
  - [ ] investigate whether travis has enough performance to do timeout tests

File: release.nix [hunk 3]
```diff
@@ -8,15 +8,17 @@ let
...
-  hevmTestReport = x: x.runCommand "hevm-test-report" {} ''
+  hevmCompliance = x: x.runCommand "hevm-compliance" {} ''
     mkdir -p $out/nix-support
     export PATH=${x.pkgs.hevm}/bin:$PATH
-    ${x.pkgs.hevm}/bin/hevm vm-test-report \
-      --tests ${ethereum-test-suite x} > $out/index.html
+    ${x.pkgs.hevm}/bin/hevm compliance \
+      --tests ${ethereum-test-suite x} \
+      --skip modexp \
+      --html > $out/index.html
```
Comment (rainbreak) [inline release.nix]: I think we don't need to save the html anymore. This was used when we had hydra, but now isn't served anywhere.

**Comment by rainbreak** [comment]
Linux tests are passing, but getting `libc++abi.dylib: terminating with uncaught exception of type int` on darwin.

File: .travis.yml [hunk 4]
```diff
@@ -1,6 +1,9 @@
 os:
   - linux
   - osx
+env:
+  - CC=gcc-5.4 CXX=g++-5.4
```
Comment (rainbreak) [inline .travis.yml]: This doesn't actually work as it isn't picked up by ghc

```
Using gcc version 4.2.1 given by user at:

/nix/store/j8k53qm6d0q8vjy7rqv007c63pn0pnnf-clang-wrapper-5.0.2/bin/clang
```

**Comment by rainbreak** [comment]
fyi @japesinator

**Comment by rainbreak** [comment]
Released as `hevm 0.33` in b206e87

**Comment by xwvvvvwx** [comment]
awesome :clap: :tada:

**Comment by mbrock** [comment]
Omg you guys 🥰

**Pull request merged**